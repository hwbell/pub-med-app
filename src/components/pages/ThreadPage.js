import React from 'react';
import '../../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import Header from '../Header';
import Thread from '../Thread';
import ThreadForm from '../ThreadForm';
import { Button } from 'reactstrap';

// functions
import { getPublicThreads, saveThread } from '../../tools/serverFunctions';

// for server requests
const user = JSON.parse(localStorage.getItem('user'));
const token = JSON.parse(localStorage.getItem('token'));
const headers = {
  Authorization: `Bearer ${token}`,
  Accept: 'application/json',
  'Content-Type': 'application/json',
}

// ******************************************************************************
class ThreadPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      sorter: '',
      showThreadForm: false,
      showUniqueWarning: false,
      threadPage: 1,
      threadSorter: 'date'
    }

    this.fetchServerThreads = this.fetchServerThreads.bind(this);
    this.renderThreads = this.renderThreads.bind(this);
    this.toggleThreadForm = this.toggleThreadForm.bind(this);
    this.handleSubmitThread = this.handleSubmitThread.bind(this);
    this.toggleUniqueWarning = this.toggleUniqueWarning.bind(this);
  }

  componentDidMount() {

    console.log(!this.props.serverThreads)

    if (!this.props.serverThreads) {
      console.log('there are no serverThreads in props')
      this.fetchServerThreads();
    }
  }

  fetchServerThreads() {
    console.log('fetching server threads');

    let { threadSorter, threadPage } = this.state;

    return getPublicThreads(headers, threadSorter, threadPage)
      .then((response) => {
        console.log(response)

        // once we have posted, we will refresh in App
        this.props.refreshServerThreads(response)


        // Fail => show the fail message / reason

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }

  // we use the isUsers flag to determine if this is the user's thread or not
  // this will allow the user to edit their own thread info, but not others (its also restricted on the backend)
  // comments are public, so any user can comment on any thread. 
  // nly thread info - name, article, paragraph - is restricted
  renderThreads(threads, isUsers) {
    let title = isUsers ? 'your threads' : 'recent threads';
    return (
      <div className="center-all-col">
        <p className="section-title">{title}</p>
        {threads.map((thread, i) => {
          return <Thread key={i}
            allowEdit={isUsers}
            thread={thread}
            handleSubmitThread={this.handleSubmitThread} 
            refreshServerThreads={this.props.refreshServerThreads}
            />
        })}
      </div>
    )
  }

  toggleUniqueWarning() {
    this.setState({
      showUniqueWarning: !this.state.showUniqueWarning
    }, () => {

      // if its now set to true, change it back after 2.4 s
      if (this.state.showUniqueWarning) {
        setTimeout(() => {
          this.setState({
            showUniqueWarning: false
          })
        }, 2400)
      }
    })
  }

  toggleThreadForm() {

    // set the selected thread if we are about to show the modal
    if (!this.state.showThreadForm) {

    }

    this.setState({
      showThreadForm: !this.state.showThreadForm
    }, () => {
      console.log(this.state.showThreadForm)
    })
  }

  // this will post our thread and relay the response back to App
  handleSubmitThread(threadInfo) {

    // we need to know if its one of the logged in user's threads. this will control
    // whether we need to update the user's threads or not
    let isUsers = threadInfo.owner === user._id; 

    // need to know if its a new thread. If it is, it won't have an owner
    let isNewThread = !!threadInfo.owner;

    // If the thread is tagged with the isComment property, it came from the CommentForm
    let isComment = threadInfo.isComment;

    // must be logged in
    if (!user || !token) {
      return;
    }

    if (!threadInfo || Object.keys(threadInfo).length === 0) {
      return;
    }
    console.log('posting thread')

    // set up post obj with user
    let post = {
      user: user.name || user.email
    };

    if (!threadInfo.isComment) {
      // Add all the non-empty entries
      Object.keys(threadInfo).forEach((key) => {
        if (threadInfo[key]) {
          post[key] = threadInfo[key]
        }
      })
    } else {
      // this means we are just posting a comment straight from the Comment Form,
      // we don't need to change it
      post = threadInfo;
    }

    // Post the thread to the server. Using the currently stored token
    // will assign the user as the owner of the thread in mongodb on the backend.
    console.log(post)
    return saveThread(post, headers, isComment)
      .then((response) => {
        console.log(response)

        // once we have posted, we will: 

        // Success => refresh the threads. This will send the new thread list sent back
        // from the server to the root App component to update all concerned components
        if (response.code !== 11000) {
          this.props.refreshServerThreads(response);

          // refresh for the user if this was a user thread, or a new thread
          if (isUsers || isNewThread) {
            this.props.refreshUserThreads(response);
          }

          if (this.state.showThreadForm) {
            this.toggleThreadForm();
          }
        } else {
          this.toggleUniqueWarning()
        }

        // Fail => show the fail message / reason

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }

  render() {

    const { user, serverThreads } = this.props;
    const haveUserThreads = user && user.threads && user.threads.length > 0;
    const haveServerThreads = serverThreads && serverThreads.length > 0;

    return (
      <div className="page">

        <ThreadForm
          toggle={this.toggleThreadForm}
          isVisible={this.state.showThreadForm}
          handleSubmitThread={this.handleSubmitThread}
          showUniqueWarning={this.state.showUniqueWarning}
          toggleUniqueWarning={this.toggleUniqueWarning}
        />

        <div className="glass page-content" >

          {/* the header */}
          <Header
            class="heading"
            title={"PMC Threads"}
            subtitle={"post a thread to start a discussion"}
          />

          {/* make a new thread */}
          <Button className="add article-button" siz="md" style={styles.button}
            onClick={this.toggleThreadForm}>start a new thread</Button>

          {haveServerThreads && this.renderThreads(serverThreads)}

          {haveUserThreads && this.renderThreads(user.threads, true)}

        </div>

      </div>
    );
  }
}

const styles = {
  button: {
    margin: '24px',
    // alignSelf: 'flex-start'
  }
}

export default ThreadPage;
