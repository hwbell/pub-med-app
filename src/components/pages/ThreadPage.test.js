import React from 'react';

import ThreadPage from './ThreadPage';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// mock server functions
jest.mock('../../tools/serverFunctions.js');

const user = {
  name: 'Mark',
  email: 'mark@example.com',
  password: 'password'
}
const token = '2e8298fgj.23d454543524.2524523542134';

const registerServerThreadsStub = jest.fn();
const fetchServerThreadsStub = jest.fn();

const refreshUserThreadsStub = jest.fn();
const renderThreadsStub = jest.fn();
const refreshServerThreadsStub = jest.fn();

const handleSubmitThreadStub = (threadInfo) => {
  return new Promise((resolve, reject) => {

    if (!threadInfo || Object.keys(threadInfo).length === 0) {
      reject('invalid input');
    } else {
      resolve(refreshUserThreadsStub());
    }
  });
}

describe('ThreadPage', () => {

  let someProps;
  let user = {};
  beforeAll(() => {
    user.threads = [
      {
        name: 'Sourcing of PMID: 013091283',
        article: '013091283',

      },
      {
        name: 'Data in figure 1 of PMID: 013091283',
        article: '013091283'
      },
      {
        name: 'Figure 3 in PMID: 013091283',
        article: '013091283'
      }
    ]
    someProps = {
      user,
      refreshUserThreads: refreshUserThreadsStub,
      refreshServerThreads: refreshServerThreadsStub,
      registerServerThreads: registerServerThreadsStub
    };
  })

  let wrapper;
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<ThreadPage {...someProps} />);
  })

  it('renders without crashing', async () => {
    wrapper.update();
  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(<ThreadPage  {...someProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the correct elements', () => {
    expect(wrapper.find('.page-content').length).toBe(1)
    expect(wrapper.find('.glass').length).toBe(1)
    expect(wrapper.find('.add').length).toBe(1);

    expect(wrapper.find('Header').length).toBe(1)

    expect(wrapper.find('Header').props().title).toBe('PMC Threads');
    expect(wrapper.find('Header').props().subtitle).toBe('post a thread to start a discussion');
  })

  it('should contain one <Thread /> for each user thread', () => {
    expect(wrapper.find('Thread').length).toBe(3);
  })

  it('should containe one <Thread/> for each server thread', () => {
    wrapper.setProps({
      serverThreads: user.threads
    });
    // now there is an additional 3
    wrapper.update();
    expect(wrapper.find('Thread').length).toBe(6);
  })

  it('should show the correct section titles for user / serverThreads', () => {
    
    // first, there's only one
    expect(wrapper.find('.section-title').length).toBe(1);
    expect(wrapper.find('.section-title').text()).toBe('your threads');
    
    wrapper.setProps({
      serverThreads: user.threads
    });
    wrapper.update();

    // now another appears with the serverThreads, but serverThreads are on top
    expect(wrapper.find('.section-title').length).toBe(2);
    expect(wrapper.find('.section-title').at(0).text()).toBe('your threads');
    expect(wrapper.find('.section-title').at(1).text()).toBe('recent threads');

  })

  it('should toggle the showThreadForm boolean', () => {

    expect(wrapper.state().showThreadForm).toBe(false);

    wrapper.find('.add').simulate('click');
    wrapper.update();
    expect(wrapper.state().showThreadForm).toBe(true);

    wrapper.find('.add').simulate('click');
    wrapper.update();
    expect(wrapper.state().showThreadForm).toBe(false);

  })

  it('should toggle the showUniqueWarning boolean', () => {

    expect(wrapper.state().showUniqueWarning).toBe(false);

    wrapper.instance().toggleUniqueWarning();
    wrapper.update();
    expect(wrapper.state().showUniqueWarning).toBe(true);
    // 
    wrapper.instance().toggleUniqueWarning();
    wrapper.update();
    expect(wrapper.state().showUniqueWarning).toBe(false);

  })

  it('should fire this.props.refreshUserThreads when this.handleSubmitThread is called', () => {

    expect(refreshUserThreadsStub.mock.calls.length).toBe(0);

    wrapper.instance().handleSubmitThread = handleSubmitThreadStub;
    wrapper.instance().handleSubmitThread({
      name: 'Sourcing of PMID: 013091283',
      article: '013091283'
    })
    wrapper.update();

    expect(refreshUserThreadsStub.mock.calls.length).toBe(1);

  })

  it('should not render Threads if there are no user threads', () => {
    someProps.user.threads = null;

    const newWrapper = shallow(<ThreadPage {...someProps} />);
    newWrapper.instance().renderThreads = renderThreadsStub;

    expect(newWrapper.find('Thread').length).toBe(0);

  })

  // if there are server collections in props, this is a result of fetching them previously and
  // passing the result back to App, which then provides them as props.
  it('should fire fetchServerThreads() on startup if there are no serverThreads as props', async () => {

    
    someProps.serverThreads = null;

    const newWrapper = shallow(<ThreadPage {...someProps} />);

    let fetchServerThreadsStub = jest.fn();
    newWrapper.instance().fetchServerThreads = fetchServerThreadsStub;
    newWrapper.instance().componentDidMount();

    expect(fetchServerThreadsStub.mock.calls.length).toBe(1)
  })

  it('should not fire fetchServerThreads() on startup if there are serverThreads as props', async () => {

    someProps.serverThreads = user.threads;
    const newWrapper = shallow(<ThreadPage {...someProps} />);

    newWrapper.instance().fetchServerThreads = fetchServerThreadsStub;

    wrapper.update();

    expect(fetchServerThreadsStub.mock.calls.length).toBe(0)
  })

  it('should fire this.props.registerServerTheads() when fetchServerThreads() is fired', async () => {
    registerServerThreadsStub.mockClear();
    
    await wrapper.instance().fetchServerThreads();
    wrapper.update();

    expect(registerServerThreadsStub.mock.calls.length).toBe(1)
  })
})