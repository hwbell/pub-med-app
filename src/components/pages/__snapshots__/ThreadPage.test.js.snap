// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadPage renders correctly 1`] = `
<div
  className="page"
>
  <div
    className="glass page-content"
  >
    <div
      className="heading"
    >
      <h2
        className="title"
      >
        PMC Threads
      </h2>
      <h4
        className="subtitle"
      >
        post a thread to start a discussion
      </h4>
    </div>
    <div
      className="outline"
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div
        className="fade"
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "position": "absolute",
            "right": "30px",
            "top": "18px",
          }
        }
      >
        <button
          aria-label={null}
          className="add article-button btn btn-secondary btn-sm"
          onClick={[Function]}
          type="button"
        >
          +new thread
        </button>
      </div>
      <div
        id="loader"
      >
        <p
          className="thread-text"
        >
          getting threads ... 
        </p>
        <div>
          <svg
            fill="whitesmoke"
            height={100}
            viewBox="0 0 120 30"
            width={100}
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="15"
              cy="15"
              r="15"
            >
              <animate
                attributeName="r"
                begin="0s"
                calcMode="linear"
                dur="0.8s"
                from="15"
                repeatCount="indefinite"
                to="15"
                values="15;9;15"
              />
              <animate
                attributeName="fillOpacity"
                begin="0s"
                calcMode="linear"
                dur="0.8s"
                from="1"
                repeatCount="indefinite"
                to="1"
                values="1;.5;1"
              />
            </circle>
            <circle
              attributeName="fillOpacity"
              cx="60"
              cy="15"
              from="1"
              r="9"
              to="0.3"
            >
              <animate
                attributeName="r"
                begin="0s"
                calcMode="linear"
                dur="0.8s"
                from="9"
                repeatCount="indefinite"
                to="9"
                values="9;15;9"
              />
              <animate
                attributeName="fillOpacity"
                begin="0s"
                calcMode="linear"
                dur="0.8s"
                from="0.5"
                repeatCount="indefinite"
                to="0.5"
                values=".5;1;.5"
              />
            </circle>
            <circle
              cx="105"
              cy="15"
              r="15"
            >
              <animate
                attributeName="r"
                begin="0s"
                calcMode="linear"
                dur="0.8s"
                from="15"
                repeatCount="indefinite"
                to="15"
                values="15;9;15"
              />
              <animate
                attributeName="fillOpacity"
                begin="0s"
                calcMode="linear"
                dur="0.8s"
                from="1"
                repeatCount="indefinite"
                to="1"
                values="1;.5;1"
              />
            </circle>
          </svg>
        </div>
      </div>
      <div
        className="center-all-col"
      >
        <p
          className="section-title"
        >
          your threads
        </p>
        <div
          className="outline"
        >
          <div
            style={
              Object {
                "alignItems": "flex-start",
                "alignSelf": "flex-start",
                "display": "flex",
                "flexDirection": "column",
                "justifyContent": "flex-start",
                "padding": "10px",
                "width": "100%",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "position": "absolute",
                  "right": "10px",
                  "top": "10px",
                }
              }
            >
              <i
                className="fas fa-edit article-button"
                onClick={[Function]}
              />
              <i
                className="fas fa-trash-alt article-button"
                onClick={[Function]}
              />
            </div>
            <p
              className="thread-title"
            >
              Sourcing of PMID: 013091283
            </p>
            <p
              className="thread-text"
            >
              <i
                className=" far fa-user"
              />
                undefined
            </p>
            <p
              className="thread-text"
            >
              @ article(s): 
              <strong>
                013091283
              </strong>
            </p>
            <hr
              style={
                Object {
                  "backgroundColor": "rgba(245,245,245,0.1)",
                  "margin": "4px",
                  "padding": 0,
                }
              }
            />
            <span
              style={
                Object {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                  "paddingTop": "20px",
                }
              }
            >
              <div>
                <p
                  className="thread-text"
                >
                  undefined comments
                </p>
              </div>
              <div
                className="collapse"
                style={Object {}}
              >
                <div
                  style={
                    Object {
                      "marginTop": "25px",
                    }
                  }
                >
                  <p
                    style={
                      Object {
                        "color": "silver",
                        "margin": "5px 20px",
                      }
                    }
                  >
                    Have something to say about this thread? Add a comment to join the discussion.
                  </p>
                  <textarea
                    className="form-control"
                    onChange={[Function]}
                    placeholder="Your comment here"
                    style={
                      Object {
                        "borderRadius": "6px",
                        "margin": "15px auto",
                        "padding": "8px",
                        "width": "95%",
                      }
                    }
                  />
                  <div
                    style={
                      Object {
                        "alignItems": "center",
                        "alignSelf": "flex-end",
                        "display": "flex",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <button
                      aria-label={null}
                      className="add article-button btn btn-secondary btn-sm"
                      onClick={[Function]}
                      type="button"
                    >
                      post comment
                    </button>
                    <button
                      aria-label={null}
                      className="back article-button btn btn-secondary btn-sm"
                      onClick={[Function]}
                      type="button"
                    >
                      cancel
                    </button>
                  </div>
                </div>
              </div>
              <div
                className="fade"
                style={
                  Object {
                    "alignSelf": "flex-end",
                  }
                }
              >
                <button
                  aria-label={null}
                  className="view article-button btn btn-secondary btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  comment
                </button>
              </div>
            </span>
          </div>
        </div>
        <div
          className="outline"
        >
          <div
            style={
              Object {
                "alignItems": "flex-start",
                "alignSelf": "flex-start",
                "display": "flex",
                "flexDirection": "column",
                "justifyContent": "flex-start",
                "padding": "10px",
                "width": "100%",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "position": "absolute",
                  "right": "10px",
                  "top": "10px",
                }
              }
            >
              <i
                className="fas fa-edit article-button"
                onClick={[Function]}
              />
              <i
                className="fas fa-trash-alt article-button"
                onClick={[Function]}
              />
            </div>
            <p
              className="thread-title"
            >
              Data in figure 1 of PMID: 013091283
            </p>
            <p
              className="thread-text"
            >
              <i
                className=" far fa-user"
              />
                undefined
            </p>
            <p
              className="thread-text"
            >
              @ article(s): 
              <strong>
                013091283
              </strong>
            </p>
            <hr
              style={
                Object {
                  "backgroundColor": "rgba(245,245,245,0.1)",
                  "margin": "4px",
                  "padding": 0,
                }
              }
            />
            <span
              style={
                Object {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                  "paddingTop": "20px",
                }
              }
            >
              <div>
                <p
                  className="thread-text"
                >
                  undefined comments
                </p>
              </div>
              <div
                className="collapse"
                style={Object {}}
              >
                <div
                  style={
                    Object {
                      "marginTop": "25px",
                    }
                  }
                >
                  <p
                    style={
                      Object {
                        "color": "silver",
                        "margin": "5px 20px",
                      }
                    }
                  >
                    Have something to say about this thread? Add a comment to join the discussion.
                  </p>
                  <textarea
                    className="form-control"
                    onChange={[Function]}
                    placeholder="Your comment here"
                    style={
                      Object {
                        "borderRadius": "6px",
                        "margin": "15px auto",
                        "padding": "8px",
                        "width": "95%",
                      }
                    }
                  />
                  <div
                    style={
                      Object {
                        "alignItems": "center",
                        "alignSelf": "flex-end",
                        "display": "flex",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <button
                      aria-label={null}
                      className="add article-button btn btn-secondary btn-sm"
                      onClick={[Function]}
                      type="button"
                    >
                      post comment
                    </button>
                    <button
                      aria-label={null}
                      className="back article-button btn btn-secondary btn-sm"
                      onClick={[Function]}
                      type="button"
                    >
                      cancel
                    </button>
                  </div>
                </div>
              </div>
              <div
                className="fade"
                style={
                  Object {
                    "alignSelf": "flex-end",
                  }
                }
              >
                <button
                  aria-label={null}
                  className="view article-button btn btn-secondary btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  comment
                </button>
              </div>
            </span>
          </div>
        </div>
        <div
          className="outline"
        >
          <div
            style={
              Object {
                "alignItems": "flex-start",
                "alignSelf": "flex-start",
                "display": "flex",
                "flexDirection": "column",
                "justifyContent": "flex-start",
                "padding": "10px",
                "width": "100%",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "position": "absolute",
                  "right": "10px",
                  "top": "10px",
                }
              }
            >
              <i
                className="fas fa-edit article-button"
                onClick={[Function]}
              />
              <i
                className="fas fa-trash-alt article-button"
                onClick={[Function]}
              />
            </div>
            <p
              className="thread-title"
            >
              Figure 3 in PMID: 013091283
            </p>
            <p
              className="thread-text"
            >
              <i
                className=" far fa-user"
              />
                undefined
            </p>
            <p
              className="thread-text"
            >
              @ article(s): 
              <strong>
                013091283
              </strong>
            </p>
            <hr
              style={
                Object {
                  "backgroundColor": "rgba(245,245,245,0.1)",
                  "margin": "4px",
                  "padding": 0,
                }
              }
            />
            <span
              style={
                Object {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                  "paddingTop": "20px",
                }
              }
            >
              <div>
                <p
                  className="thread-text"
                >
                  undefined comments
                </p>
              </div>
              <div
                className="collapse"
                style={Object {}}
              >
                <div
                  style={
                    Object {
                      "marginTop": "25px",
                    }
                  }
                >
                  <p
                    style={
                      Object {
                        "color": "silver",
                        "margin": "5px 20px",
                      }
                    }
                  >
                    Have something to say about this thread? Add a comment to join the discussion.
                  </p>
                  <textarea
                    className="form-control"
                    onChange={[Function]}
                    placeholder="Your comment here"
                    style={
                      Object {
                        "borderRadius": "6px",
                        "margin": "15px auto",
                        "padding": "8px",
                        "width": "95%",
                      }
                    }
                  />
                  <div
                    style={
                      Object {
                        "alignItems": "center",
                        "alignSelf": "flex-end",
                        "display": "flex",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <button
                      aria-label={null}
                      className="add article-button btn btn-secondary btn-sm"
                      onClick={[Function]}
                      type="button"
                    >
                      post comment
                    </button>
                    <button
                      aria-label={null}
                      className="back article-button btn btn-secondary btn-sm"
                      onClick={[Function]}
                      type="button"
                    >
                      cancel
                    </button>
                  </div>
                </div>
              </div>
              <div
                className="fade"
                style={
                  Object {
                    "alignSelf": "flex-end",
                  }
                }
              >
                <button
                  aria-label={null}
                  className="view article-button btn btn-secondary btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  comment
                </button>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
